enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -183,17 +185,19 @@ func (h *HitTrackerFactory) enqueue(ctx context.Context, requestMetadata *repb.R
 		}
 		return
 	}
-	defer h.mu.Unlock()
 
-	groupID := h.groupID(ctx)
 	if _, ok := h.hitsByGroup[groupID]; !ok {
 		hits := cacheHits{gid: groupID, hits: []*hitpb.CacheHit{}}
 		h.hitsByGroup[groupID] = &hits
 		h.hitsQueue = append(h.hitsQueue, &hits)
 	}
 
 	groupHits := h.hitsByGroup[groupID]
+	h.mu.Unlock()
+	authHeaders := authutil.GetAuthHeaders(ctx, h.authenticator)
+	groupHits.mu.Lock()
 	if len(groupHits.hits) >= h.maxPendingHitsPerGroup {
+		groupHits.mu.Unlock()
```

Comment:
nit: could add a method `func (ch *cacheHits) Add(hit *hitpbCache.Hit) (ok bool)` which would let you unlock w/ `defer`. (If `ok` returns false then you could alert/update metrics/return)